Sources/TestsCommon/CommonTestUtils.swift

```diff
@@ -139,21 +139,45 @@ open class MongoSwiftTestCase: XCTestCase {
 public enum TestTopologyConfiguration: String, Decodable {
     case sharded
     case replicaSet = "replicaset"
+    case shardedReplicaSet = "sharded-replicaset"
     case single
 
-    /// Determines the topologyType of a client based on the reply returned by running an isMaster command
-    public init(isMasterReply: BSONDocument) throws {
+    /// Determines the topologyType of a client based on the reply returned by running an isMaster command and the
+    /// first document in the config.shards collection.
+    public init(isMasterReply: BSONDocument, shards: [BSONDocument]) throws {
         // Check for symptoms of different topologies
         if isMasterReply["msg"] != "isdbgrid" &&
             isMasterReply["setName"] == nil &&
             isMasterReply["isreplicaset"] != true {
             self = .single
         } else if isMasterReply["msg"] == "isdbgrid" {
-            self = .sharded
+            guard !shards.isEmpty else {
+                throw TestError(
+                    message: "isMasterReply \(isMasterReply) implies a sharded cluster, but config.shards is empty"
+                )
+            }
+            for shard in shards {
+                guard let host = shard["host"]?.stringValue else {
+                    throw TestError(message: "config.shards document \(shard) unexpectedly missing host string")
+                }
+                // If the shard is backed by a single server, this field will contain a single host (e.g. localhost:27017).
+                // If the shard is backed by a replica set, this field will contain the name of the replica followed by a
+                // forward slash and a comma-delimited list of hosts.
+                let replSetHostRegex = try NSRegularExpression(pattern: #"^.*\/.*:\d+$"#)
```

Comment:
inspecting documents in config.shards is the way [the spec](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#determining-if-a-sharded-cluster-uses-replica-sets) instructs you to check for this.

borrowed this regex approach from [PHP](https://github.com/mongodb/mongo-php-library/blob/6c66a1a291f4309e910e1c29998550aa971de220/tests/FunctionalTestCase.php#L333).

I'm not sure if it's possible to have a sharded cluster that is backed by a mix of standalones and replica sets, but for testing purposes it seems to me that we shouldn't count it as being fully backed by replica sets, so we'll mark it `.sharded` here if that is the case.